components/modules/otahttp.c

```diff
@@ -0,0 +1,205 @@
+// Module for downloading over-the-air (OTA) firmware upgrades via http
+
+#include <string.h>
+#include <sys/socket.h>
+#include <netdb.h>
+#include <stdarg.h>
+
+#include "module.h"
+#include "lauxlib.h"
+#include "platform.h"
+
+#include "esp_system.h"
+#include "esp_log.h"
+#include "esp_ota_ops.h"
+#include "lwip/dns.h"
+
+#include "ota_common.h"
+
+#define BUFFSIZE 1024
+#define TEXT_BUFFSIZE 1024
+
+static const char *TAG = "otahttp";
+
+#define OTAHTTP_ERROR(L, format,...) \
+  ESP_LOGE(TAG, format, ##__VA_ARGS__); \
+  lua_pushboolean (L, false); \
+  return 1;
+
+typedef struct {
+  bool completed;
+  ip_addr_t ip_address;
+} otahttp_ip_lookup;
+
+static void dns_lookup_callback(const char *name, const ip_addr_t *ipaddr, void *callback_arg)
+{
+    otahttp_ip_lookup* lookup = callback_arg;
+    lookup->ip_address = *ipaddr;
+    lookup->completed = true;
+}
+
+static ip_addr_t dns_lookup(const char *domain) {
+  otahttp_ip_lookup lookup;
+
+  lookup.completed = false;
+  IP_ADDR4( &(lookup.ip_address), 0,0,0,0 );
+
+  err_t  err = dns_gethostbyname(domain, &(lookup.ip_address), dns_lookup_callback, &lookup);
+  if(err == ERR_OK)
+    dns_lookup_callback(domain, &(lookup.ip_address), &lookup);
+  else if (err == ERR_INPROGRESS)
+    while(!lookup.completed);
```

Comment:
This hard loop here isn't good, nor the lack of synchronisation on the `completed` flag between threads. There are multiple constructs available to suspend a task which would be far better. The canonical way of doing it is properly async. Have a look at the `net` module for inspiration.